### PR TITLE
Ian UI d2 1394 update attest failure reaction

### PR DESCRIPTION
**Pull request by Ian-Nara**
Update shared to inform attestation response watcher of both status code and response body.

### CODE REVIEW

File: src/main/java/com/uid2/shared/attest/UidCoreClient.java
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/attest/UidCoreClient.java:120]: Why did you remove this?

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java:37]: Are you sure this is only used by operator?

File: src/main/java/com/uid2/shared/attest/UidCoreClient.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/UidCoreClient.java:120]: We want to decouple the attest() calls from the calls to a refresh endpoint.

"The renewal of the attestation process should be changed to be done as a background process, rather than being done as part of a request - response call from a client of the Operator. This would allow for the retries to not block an incoming request."

Calls to attest are now call handled within the AttestationTokenRetreiver class

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java:37]: Out of the core, optout, admin, and operator, the attestation token retriever is used by operator and optout, but optout uses a null responseWatcher. Will look into optout closer to make sure it is not broken by these changes

**Review by zaiweidu — COMMENTED** [review]
LGTM

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java:37]: I have tested optout/core with my shared changes locally and it is able to attest and refresh with no extra changes needed

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (thomasm-ttd) [inline src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java:99]: Why have you moved this to the constructor? Currently it is only set after the first attestation, but now it will happen 60s after the construction, regardless of whether attest has been called.

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (thomasm-ttd) [inline src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java:115]: Why have you added this check and then attest? The scheduled attestationCheck should only happen after the first successful attestation

File: src/main/java/com/uid2/shared/attest/UidCoreClient.java
Comment (thomasm-ttd) [reply src/main/java/com/uid2/shared/attest/UidCoreClient.java:120]: This work was already done, and attestation now happens 10 minutes before the expiry of the token. I don't believe we need to change this. getWithAttest allows the first request to perform attestation, at which point the schedule will start.

File: src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java
Comment (thomasm-ttd) [inline src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java:93]: Why do we have 3 401 responses with 'bad' body and 3 401 responses  with the more detailed response?

File: src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java
Comment (thomasm-ttd) [inline src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java:76]: Good test

File: src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java
Comment (thomasm-ttd) [inline src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java:91]: maybe we should reduce the retry interval for the test. This test will take 3 seconds every time

**Review by thomasm-ttd — COMMENTED** [review]
I don't believe we need to change the attestation scheduled behaviour. Attestation happens on the first call to core, and then after that 10 minutes before expiry of the token.

File: src/main/java/com/uid2/shared/attest/UidCoreClient.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/UidCoreClient.java]: Ah I see I was not considering that the expiration check will handle the retry already for failure in all cases except if the first attestation call fails. I have undone this change.

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java]: changed back

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java]: changed back as noted above.

File: src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java
Comment (Ian-Nara) [reply src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java:78]: reduced the interval

File: src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java
Comment (Ian-Nara) [reply src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java]: modified this test. It now has an intial sucess followed by 2 failures, followed by a sucess

**Review by thomasm-ttd — APPROVED** [review]

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (thomasm-ttd) [inline src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java:118]: Extra line

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (thomasm-ttd) [inline src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java:133]: I know this was there before, but what would cause the IOException? Should we be returning a 401 for an IOException? Or is it a 500?

Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java]: httpClient.post() is what throws it. Changed this to return a 500

Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java]: Changed this to a 500 for an IOException, and fixed an issue in URLConnectionHttpClient.java that resulted in IOException being thrown for any non-200 status code

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (thomasm-ttd) [inline src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java]: Sorry to do this as I know you have been trying to get this approved for a while, but can we log out for both the 401 and the 500 the response code, and change the messages so we can tell what the cause was?

File: src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java
Comment (thomasm-ttd) [inline src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java:90]: We should use 500 as the error code here. If we use 401, it implies the operator will retry on a 401, which it won't, it will shutdown

**Review by thomasm-ttd — COMMENTED** [review]
This will break the build for operator, so please make sure you are ready to merge operator soon after merging this change, publishing it and updating operator

File: src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java]: I am not quite sure what you are suggesting. It will already log out for both a 401 and a 500, with the message being based on the exception that occurred. Are you suggesting we generalize the error that is logged, so that a 401 always logs the same thing and a 500 always logs the same thing? And if so what do you suggest for the message?

Comment (thomasm-ttd) [reply src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java]: What I want to make sure is that we can tell the difference in our logs (so we can alert differently) between the 2 types of 401 errors. If you are confident that this is the case, then I am OK with this.

**Review by thomasm-ttd — APPROVED** [review]
Please check that all the conversations have been resolved before merging

**Pull request merged**
